k4FWCore/components/CollectionMerger.cpp

```diff
@@ -0,0 +1,123 @@
+        &CollectionMerger::mergeCollections<edm4hep::MCRecoParticleAssociationCollection>;
+    m_map["edm4hep::MCRecoCaloAssociationCollection"] =
+        &CollectionMerger::mergeCollections<edm4hep::MCRecoCaloAssociationCollection>;
+    m_map["edm4hep::MCRecoTrackerAssociationCollection"] =
+        &CollectionMerger::mergeCollections<edm4hep::MCRecoTrackerAssociationCollection>;
+    m_map["edm4hep::MCRecoCaloParticleAssociationCollection"] =
+        &CollectionMerger::mergeCollections<edm4hep::MCRecoCaloParticleAssociationCollection>;
+    m_map["edm4hep::MCRecoClusterParticleAssociationCollection"] =
+        &CollectionMerger::mergeCollections<edm4hep::MCRecoClusterParticleAssociationCollection>;
+    m_map["edm4hep::MCRecoTrackParticleAssociationCollection"] =
+        &CollectionMerger::mergeCollections<edm4hep::MCRecoTrackParticleAssociationCollection>;
+    m_map["edm4hep::RecoParticleVertexAssociationCollection"] =
+        &CollectionMerger::mergeCollections<edm4hep::RecoParticleVertexAssociationCollection>;
+    m_map["edm4hep::TimeSeriesCollection"] = &CollectionMerger::mergeCollections<edm4hep::TimeSeriesCollection>;
+    m_map["edm4hep::RecDqdxCollection"]    = &CollectionMerger::mergeCollections<edm4hep::RecDqdxCollection>;
+  }
+
+  std::shared_ptr<podio::CollectionBase> operator()(
+      const std::map<std::string, const std::shared_ptr<podio::CollectionBase>&>& input) const override {
+    std::shared_ptr<podio::CollectionBase> ret;
+    debug() << "Merging " << input.size() << " collections" << endmsg;
+    std::string_view type = "";
+    for (const auto& [name, coll] : input) {
+      debug() << "Merging collection " << name << " of type " << coll->getTypeName() << " with " << coll->size()
+              << " elements" << endmsg;
+      if (type.empty()) {
+        type = coll->getTypeName();
+      } else if (type != coll->getTypeName()) {
+        throw std::runtime_error("Different collection types are not supported");
+        return ret;
+      }
+      (this->*m_map.at(coll->getTypeName()))(coll, ret);
+    }
+    return ret;
+  }
+
+private:
+  using MergeType = void (CollectionMerger::*)(const std::shared_ptr<podio::CollectionBase>&,
+                                               std::shared_ptr<podio::CollectionBase>&) const;
+  std::map<std::string_view, MergeType> m_map;
+  Gaudi::Property<bool>                 m_copy{this, "Copy", false,
+                               "Copy the elements of the collections instead of creating a subset collection"};
+
+  template <typename T>
+  void mergeCollections(const std::shared_ptr<podio::CollectionBase>& source,
+                        std::shared_ptr<podio::CollectionBase>&       ret) const {
+    if (!ret) {
+      ret.reset(new T());
+      if (!m_copy) {
+        ret->setSubsetCollection();
+      }
+    }
+    const auto ptr        = std::dynamic_pointer_cast<T>(ret);
+    const auto sourceColl = std::dynamic_pointer_cast<T>(source);
```

Comment:
If we `dynamic_cast` we should also check the results.

Reply:
These should be static, changed. There should never be any confusion as long as the type one gets with `getTypeName` is the same as the actual type.